cypress-io/cypress
# feat: IATR cloud data backend

- Closes https://github.com/cypress-io/cypress/issues/24440

### User facing changelog

N/A merging to feature branch

### Additional details

**Background**
This PR moves the logic for determining the "relevant" run for the new Debug page to the server and exposing it through a GraphQL query.  The server handles fetching the git hashes to send to the Cypress cloud and processing the runs returned to determine the appropriate run(s) for the front end.

**What is a "relevant" run?**
A "relevant" run for the Debug page is determined by looking up runs in a Cypress Cloud project that have a Git commit sha that matches one of the last 100 commit shas in the users's current development project.  If the current environment does not use Git, then the Debug feature will not be available for that project.  See the `RelevantRunsDataSource` to see the logic for how the "current" and "next" relevant runs are determined.

**Main Components**
* packages/data-context/src/sources/RelevantRunsDatasource.ts
  * Server data source that determines which CloudRuns match the local Git shas and should be used for the Debug page
* packages/data-context/src/sources/RelevantRunSoecsDatasource.ts
  * Server data source that pulls spec counts for the current and next runs.  Used by a poller that backs a new GQL subscription 
* packages/app/src/composables/useRelevantRun.ts
  * GQL query for querying for the relevant run order to use the runNumber in subsequent queries
* packages/app/src/composables/useTestsForDebug.ts
  * GQL query for finding the failed tests to use for a filter in the runner when filtering in debug mode 

**GQL Updates/Additions**
* CurrentProject
  * added `testsForReviewBySpec`
  * TODO: Need to complete resolver for `testsForReviewBySpec`.  Currently just a placeholder 
* Mutation
  * added `movetoNextRelevantRun` for front end to indicate that the cached run just be removed and the newest run used as the current run
* RelevantRun
  * New field type added for representing the relevant current and next run
* Subscription
  * added `relevantRuns` to allow for the front end to see changes to the current and next relevant runs
  * added `relevantRunSpecChange` to allow for the banner and page that are showing updating spec counts for RUNNING builds to update

**Dependency Updates**
* Added `mocha-junit-reporter` and `moch-multi-reporters` to the `data-context` package directly
  * This was required to fix a bug when a test would fail an assertion.  These reporters were being referenced from the root `package.json`.  They would make a call out to the peer dependency for `mocha` which is currently an older version at the root of the project. Various versions of mocha are used throughout the mono-repo that should be looked at. 

**Other Updates**
Updated the various parts of the Debug feature that was using hardcoded `runNumber`s to use the new `useRelevantRuns` composable and delay the initial queries until the value is found.

### Steps to test


_In order to test this PR, you will currently need to have the Cypress Cloud running locally or point to Cloud staging_

* Open Cypress and log in to your local Cloud instance
* Open a project that has recorded runs
* Verify that the Debug page and the sidebar badge represent the appropriate "relevant" run based on the algorithm defined above and documented in the code in `RelevantRunsDataSource`
* Start recording a new run to this project
* Verify that the new RUNNING run is shown in the banner on the Debug page as running (banner implemented here https://github.com/cypress-io/cypress/issues/24853)
* When the run completes, verify that the banner updates to show the completed status and a "View run" link (banner implemented here https://github.com/cypress-io/cypress/issues/24853)
* Click the "View run" and verify that the latest run is now shown as the current run on the Debug page



### How has the user experience changed?


N/A Backend changes only

### PR Tasks
